x/market/handler/handler_test.go

```diff
@@ -0,0 +1,203 @@
+package handler_test
+
+import (
+	"math"
+	"testing"
+
+	"github.com/cosmos/cosmos-sdk/store"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/bank"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/rand"
+	dbm "github.com/tendermint/tm-db"
+
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/ovrclk/akash/app"
+	"github.com/ovrclk/akash/testutil"
+	dkeeper "github.com/ovrclk/akash/x/deployment/keeper"
+	dtypes "github.com/ovrclk/akash/x/deployment/types"
+	"github.com/ovrclk/akash/x/market/handler"
+	"github.com/ovrclk/akash/x/market/keeper"
+	"github.com/ovrclk/akash/x/market/types"
+	pkeeper "github.com/ovrclk/akash/x/provider/keeper"
+	ptypes "github.com/ovrclk/akash/x/provider/types"
+)
+
+type testSuite struct {
+	t       *testing.T
+	ms      sdk.CommitMultiStore
+	ctx     sdk.Context
+	mkeeper keeper.Keeper
+	dkeeper dkeeper.Keeper
+	pkeeper pkeeper.Keeper
+	bkeeper bank.Keeper
+
+	handler sdk.Handler
+}
+
+func setupTestContext(t *testing.T) *testSuite {
+	suite := &testSuite{
+		t: t,
+	}
+
+	testutil.AccAddress(t)
+	mKey := sdk.NewKVStoreKey(types.StoreKey)
+	dKey := sdk.NewKVStoreKey(dtypes.StoreKey)
+	pKey := sdk.NewKVStoreKey(ptypes.StoreKey)
+
+	db := dbm.NewMemDB()
+	suite.ms = store.NewCommitMultiStore(db)
+	suite.ms.MountStoreWithDB(mKey, sdk.StoreTypeIAVL, db)
+	suite.ms.MountStoreWithDB(dKey, sdk.StoreTypeIAVL, db)
+	suite.ms.MountStoreWithDB(pKey, sdk.StoreTypeIAVL, db)
+
+	err := suite.ms.LoadLatestVersion()
+	require.NoError(t, err)
+
+	suite.ctx = sdk.NewContext(suite.ms, abci.Header{}, true, log.NewNopLogger())
+
+	suite.mkeeper = keeper.NewKeeper(app.MakeCodec(), dKey)
+	suite.dkeeper = dkeeper.NewKeeper(app.MakeCodec(), mKey)
+	suite.pkeeper = pkeeper.NewKeeper(app.MakeCodec(), pKey)
+
+	suite.handler = handler.NewHandler(handler.Keepers{
+		Market:     suite.mkeeper,
+		Deployment: suite.dkeeper,
+		Provider:   suite.pkeeper,
+		Bank:       suite.bkeeper,
+	})
+
+	return suite
+}
+
+func TestCreateBidValid(t *testing.T) {
+	suite := setupTestContext(t)
+
+	order, gspec := suite.createOrder(nil)
+
+	msg := types.MsgCreateBid{
+		Order:    order.ID(),
+		Provider: suite.createProvider(gspec.Requirements).Owner,
+		Price:    sdk.Coin{},
```

Comment:
It should error if the price is invalid (wrong denomination, too high (you check this later), etc...)